includes/Vendor/Coupon.php

```diff
@@ -0,0 +1,313 @@
+
+        if ( isset( $item_wise_discounts[ $coupon->get_code() ][ $item_key ] ) ) {
+            $discount_amount = $item_wise_discounts[ $coupon->get_code() ][ $item_key ];
+
+            $coupon_info[ $coupon->get_code() ] = [
+                'discount'    => $discount_amount,
+                'coupon_code' => $coupon->get_code(),
+                'per_qty_amount'    => $item_object['quantity'] > 0 ? ( $discount_amount / $item_object['quantity'] ) : 0,
+                'quantity'  => $item_object['quantity'],
+            ];
+        }
+
+        $item_object[ self::DOKAN_COUPON_META_KEY ] = $coupon_info;
+        $cart->cart_contents[ $item_object['key'] ] = $item_object;
+    }
+
+    /**
+     * Save coupon data to an order item.
+     *
+     * @param WC_Coupon $coupon The coupon being applied.
+     * @param WC_Discounts $discounts Discount object.
+     * @param object $item The order item object.
+     * @param WC_Order $order The order object.
+     *
+     * @return void
+     * @throws Exception
+     */
+    protected function save_coupon_data_to_order_item( WC_Coupon $coupon, WC_Discounts $discounts, $item, WC_Order $order ): void {
+        $coupon_codes = $order->get_coupon_codes();
+
+        /** @var WC_Order_Item_Product $order_item */
+        $order_item = $item->object;
+        $item_key = $order_item->get_id();
+
+        $coupon_info = $order_item->get_meta( self::DOKAN_COUPON_META_KEY );
+
+        if ( ! is_array( $coupon_info ) ) {
+            $coupon_info = [];
+        }
+
+        $coupon_info = array_filter(
+            $coupon_info,
+            function ( $coupon_temp ) use ( $coupon_codes ) {
+                return in_array( $coupon_temp['coupon_code'], $coupon_codes, true );
+            }
+        );
+
+        $item_wise_discounts = $discounts->get_discounts();
+
+        if ( isset( $item_wise_discounts[ $coupon->get_code() ][ $item_key ] ) ) {
+            $discount_amount = $item_wise_discounts[ $coupon->get_code() ][ $item_key ];
+
+            $coupon_info[ $coupon->get_code() ] = [
+                'discount' => $discount_amount,
+                'coupon_code' => $coupon->get_code(),
+                'per_qty_amount' => $order_item->get_quantity() > 0 ? $discount_amount / $order_item->get_quantity() : 0,
+                'quantity' => $order_item->get_quantity(),
+            ];
+        }
+
+        wc_update_order_item_meta( $order_item->get_id(), self::DOKAN_COUPON_META_KEY, $coupon_info );
+
+        // apply coupon sub order
+        if ( $order->get_meta( 'has_sub_order' ) ) {
+            $this->process_coupon_into_child_orders( $order, $coupon );
+        }
+    }
+
+    /**
+    * Process coupon for child orders.
+    *
+    * @param WC_Order $order
+    * @param WC_Coupon $coupon
+    *
+    * @return void
+    * @throws Exception
+    */
+    public function process_coupon_into_child_orders( WC_Order $order, WC_Coupon $coupon ): void {
```

Comment:
- Rename the method to `apply_coupon_to_child_orders`
- Change the visibility to protected if it is not used by the others.